// eify/dist/es/AbstractEventEmitter.js
import { __assign, __awaiter, __generator, __spreadArray } from "tslib";
var AbstractEventEmitter = (function () {
    function AbstractEventEmitter(options) {
        var _a = __assign({}, options), _b = _a.maxListeners, maxListeners = _b === void 0 ? 1024 : _b, _c = _a.requireLoopRemove, requireLoopRemove = _c === void 0 ? false : _c;
        this.__maxListeners = maxListeners;
        this.__listeners = {};
        this.__requireLoopRemove = requireLoopRemove;
    }
    AbstractEventEmitter.prototype.addListener = function (name, listener) {
        this.__listeners[name] = this.__listeners[name] || [];
        this.__listeners[name].push(listener);
        var maxListeners = this.__maxListeners;
        if (this.__listeners[name].length > maxListeners) {
            var error = "The '".concat(name, "' listener is not more than ").concat(maxListeners);
            console.warn(error, this);
        }
    };
    AbstractEventEmitter.prototype.on = function (name, listener) {
        return this.addListener(name, listener);
    };
    AbstractEventEmitter.prototype.removeListener = function (name, listener) {
        if (!this.__listeners[name] || !listener)
            return;
        this.__listeners[name] = this.__listeners[name].filter(function (it) { return it !== listener; });
    };
    AbstractEventEmitter.prototype.off = function (name, listener) {
        return this.removeListener(name, listener);
    };
    AbstractEventEmitter.prototype.removeAllListener = function (name) {
        var _this = this;
        if (name) {
            if (this.__requireLoopRemove && this.__listeners[name]) {
                this.__listeners[name].slice(0).forEach(function (it) {
                    _this.removeListener(name, it);
                });
            }
            this.__listeners[name] = [];
        }
        else {
            if (this.__requireLoopRemove) {
                Object.keys(this.__listeners).forEach(function (it) {
                    _this.removeAllListener(it);
                });
            }
            this.__listeners = {};
        }
    };
    AbstractEventEmitter.prototype.emit = function (name) {
        var _this = this;
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        if (!this.__listeners[name])
            return;
        var listeners = this.__listeners[name].slice(0);
        return listeners.map(function (it) { return it.call.apply(it, __spreadArray([_this], args, false)); });
    };
    AbstractEventEmitter.prototype.emitSerial = function (name) {
        var _this = this;
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        if (!this.__listeners[name])
            return;
        var listeners = this.__listeners[name].slice(0);
        return listeners.reduce(function (result, it) { return __awaiter(_this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, result];
                    case 1:
                        if ((_a.sent()) === false)
                            return [2, false];
                        return [2, it.call.apply(it, __spreadArray([this], args, false))];
                }
            });
        }); }, true);
    };
    AbstractEventEmitter.prototype.emitParallel = function (name) {
        var _this = this;
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        if (!this.__listeners[name])
            return;
        var listeners = this.__listeners[name].slice(0);
        return Promise.all(listeners.map(function (it) { return it.call.apply(it, __spreadArray([_this], args, false)); }));
    };
    AbstractEventEmitter.prototype.emitAsync = function (name) {
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        return this.emitParallel.apply(this, __spreadArray([name], args, false));
    };
    return AbstractEventEmitter;
}());
export { AbstractEventEmitter };
